fix online status indicator

Lumiphare opened the pull request:
#10005 

The original online status query simply checked whether the rendezvous_server contained the given ID. After the modification, the server address is obtained from the original set of IDs, and then a socket connection is established with that server in order to query the online status.

By using a HashMap, the same server address only needs one socket connection to send a batch of IDs at once. In the HashMap’s key, “rendezvous_server” is used to store cases where there is no “@” in the ID (i.e., the original situation where IDs only have an ID). “public” is used when the string after “@” is “public”. In other scenarios, the string after “@” is treated as a domain name or IP address (which can be IPv4 or IPv6, with optional ports; if a port is provided with an IPv6 address, it should be enclosed in brackets; no brackets are needed if there is no port). The value in the HashMap contains all IDs that need to be queried on the server corresponding to that key.

Originally, config.rs’s get_rendezvous_server function simply checked for the presence of “:” to determine if there was a port, ignoring IPv6. Therefore, IPv6 detection has been added by counting the number of “:” to handle IPv6 addresses.

I’ve tested the scenario between two Windows 10 machines myself, covering ID, ID@public, ID@IPv4 , ID@IPv4 :port, ID@IPv6, and ID@[IPv6]:port. The “online indicator” lights up as expected in all cases, but turning it off responds a bit slowly

![PixPin_2025-01-01_06-54-39](https://github.com/user-attachments/assets/97ef2a53-df23-44ee-9e1b-7fd3fbd2f785)

rustdesk commented on libs/hbb_common/src/config.rs:
```diff
@@ -736,10 +736,17 @@ impl Config {
+            1 => rendezvous_server,
+            _ => {
+                if rendezvous_server.starts_with('[') {
+                    format!("{rendezvous_server}:{RENDEZVOUS_PORT}")
```
Starting with `[` means it has ending port, why append again?

Lumiphare replied:
Sorry, that was my oversight. I've replaced it with the already implemented function as an alternative.

rustdesk commented on src/client.rs:
```diff
@@ -3516,82 +3516,121 @@ pub mod peer_online {
+    async fn create_online_stream(server_addr: String) -> ResultType<FramedStream> {
+        let (rendezvous_server, _servers, _contained) = match server_addr.as_str() {
+            "public" => {
+                (format!("{}:{}",hbb_common::config::RENDEZVOUS_SERVERS[0],hbb_common::config::RENDEZVOUS_PORT), vec![], false)
```
format?

rustdesk commented on src/client.rs:
```diff
@@ -3516,82 +3516,121 @@ pub mod peer_online {
+        let port;
+        let online_server;
+        if let Some(pos) = rendezvous_server.rfind(':') {
+            port = rendezvous_server[pos + 1 ..].parse().unwrap_or(0);
```
How about if rendezvous_server ends with ':'? What will happen to [pos + 1 ..]?

rustdesk commented on src/client.rs:
```diff
@@ -3516,82 +3516,121 @@ pub mod peer_online {
         connect_tcp(online_server, CONNECT_TIMEOUT).await
     }
 
     async fn query_online_states_(
```
I would suggest you create another function 
```
query_online_states_with_ip_port(...) 
```
to wrap old query_online_states_, so that we have minimum code change.

Commit:
Increase concurrent polling to query online states

Commit:
Prevent possible out-of-bounds access with array indices

Lumiphare commented:
I found that this issue isn’t that easy to solve properly. The previous implementation was too rough, so I spent a bit longer rewriting it. There may still be some issues.


I moved the part I implemented myself into a new function. Previously it was called `ip_port,` which was not comprehensive because it didn’t include `public` or domain names, so I changed it to `server_addr.` Most of the logic has been moved to the new `query_online_states_with_server_addr` function. I changed the function to a concurrent + polling mode. `query_online_states_with_server_addr` primarily uses the current record `ONLINE_IDS` to return user status, while user status updates are handled by each polling task. Therefore, based on the incoming  `ids`, the function also creates polling tasks. It uses `RUNNING_POLLING` to keep track of each async task, whose value is an `AtomicBool` used to signal the task to stop. Currently, the polling interval is set to 5 seconds. `FIRST_CALL` is mainly for the first load so that a portion of the user status can be displayed quickly. Since the `CONNECT_TIMEOUT` is set to 18,000 ms, whether a user goes from online to offline or offline to online, there might be situations where an offline state needs a long query time. Hence, user status changes won’t be reflected immediately.


Besides the changes I made in the new function, in order to implement concurrency, I modified the original ``#[tokio::main(flavor = "current_thread")]`` in `start_flutter_async_runner_` to **#[tokio::main(flavor = "multi_thread")]** and also updated `check_port` in `config.rs`.


For the issue at `pos + 1,` previously I considered that each `server_addr` call would run through the `check_port` function, which automatically appends a port, so it’s unlikely to end with `:`. But to be safe, I still added a check.


There may still be some parts of the code that need further changes. I’ll go ahead and submit this PR so we can revise it afterward.

Commit:
Resolve the issue where the socket cannot be reused for a persistent connection, and modify it so that it becomes a long-running connection.

Lumiphare commented:
I've now switched to a persistent connection mode. Each time, I first attempt to retrieve a `FramedStream` from `ONLINE_SOCKETS`; if it isn't found, a new socket is created. The original code was unable to establish a persistent connection because on the server side, after calling `next_timeout()`, the socket was immediately dropped—no loop was used, so each socket could only be used once.  I've changed it to a loop-based approach. This is the PR I submitted on the server side ([Solve the issue of reusing the client's socket for multiple send operations. #498](https://github.com/rustdesk/rustdesk-server/pull/498))

@rustdesk Are there any other parts that need to be modified?
